code/cli/src/crowdloan.rs

```diff
@@ -0,0 +1,215 @@
+use sp_runtime::AccountId32;
+use ss58_registry::Ss58AddressFormatRegistry;
+use std::collections::BTreeMap;
+use substrate_api_client::{
+	compose_call, compose_extrinsic, rpc::WsRpcClient, Api, AssetTipExtrinsicParams,
+	GenericAddress, XtStatus,
+};
+
+use crate::common::api_wrap;
+
+const REWARDS_PERMALINK: &str = "https://raw.githubusercontent.com/ComposableFi/composable/f887b58833f8a1ea82cdf0ffd93e3848174a5185/composablejs/packages/bootstrap_pallets/src/constants/rewards.json";
+const REWARDS_DIGEST_SHA256: &str =
+	"69e2fb1b8fbbeb0028c6c841c744a7c5744c06c4947a8673d8c9f6198ab458bb";
+const REWARDS_VESTING_PERIOD_48_WEEKS: u64 = 29030400000;
+// Submit 1000 rewards
+const REWARDS_BATCH_POPULATE_SIZE: usize = 2000;
+const REWARDS_TOTAL_EXPECTED: u128 = 3000000000000000009314;
+const REWARDS_TOTAL_CONTRIBUTOR_EXPECTED: usize = 10685;
+// Kusama
+const RELAYCHAIN_ADDRESS_FORMAT: Ss58AddressFormatRegistry =
+	Ss58AddressFormatRegistry::KusamaAccount;
+const PARACHAIN_DECIMALS: u32 = 12;
+
+type CanonicalContributorAccount = RemoteAccount<AccountId32>;
+type CanonicalContributorAmount = u128;
+
+type RawContributorAccount = String;
+type RawContributorReward = String;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Rewards<K: Ord, V> {
+	#[serde(flatten)]
+	contributors: BTreeMap<K, V>,
+}
+
+#[derive(Debug)]
+pub enum CrowdloanError {
+	UnableToDownloadContributorFile,
+	UnableToDecodeContributorFile,
+	CannotDeserializeContributorFile,
+	InvalidContributorAddress,
+	InvalidContributorReward,
+	InvalidContributorAddressPrefix,
+	ApiFailure(substrate_api_client::ApiClientError),
+}
+
+impl From<substrate_api_client::ApiClientError> for CrowdloanError {
+	fn from(x: substrate_api_client::ApiClientError) -> Self {
+		Self::ApiFailure(x)
+	}
+}
+
+fn reward_canonicalize(
+	(account, reward): (RawContributorAccount, RawContributorReward),
+) -> Result<(CanonicalContributorAccount, CanonicalContributorAmount), CrowdloanError> {
+	let reward = (str::parse::<f64>(&reward)
+		.map_err(|_| CrowdloanError::InvalidContributorReward)? *
+		10f64.powi(PARACHAIN_DECIMALS as i32)) as u128;
+	match account.strip_prefix("0x") {
+		Some(ethereum_address_src) => {
+			let ethereum_address = TryInto::<[u8; 20]>::try_into(
+				hex::decode(ethereum_address_src)
+					.map_err(|_| CrowdloanError::InvalidContributorAddress)?,
+			)
+			.map_err(|_| CrowdloanError::InvalidContributorAddress)?;
+			Ok((RemoteAccount::Ethereum(EthereumAddress(ethereum_address)), reward))
+		},
+		None => {
+			let (relaychain_account, ss58_address_format) =
+				Ss58Codec::from_ss58check_with_version(&account)
+					.map_err(|_| CrowdloanError::InvalidContributorAddress)?;
+			// Contributor are either eth or from ksm
+			if ss58_address_format != RELAYCHAIN_ADDRESS_FORMAT.into() {
+				Err(CrowdloanError::InvalidContributorAddressPrefix)
+			} else {
+				Ok((RemoteAccount::RelayChain(relaychain_account), reward))
+			}
+		},
+	}
+}
+
+pub fn crowdloan_seed(
```

Comment:
Would be nice to log the time taken to execute for each action

Reply:
Probably another PR